app/src/main/java/com/epfl/drawyourpath/community/CommunityTournamentPostViewAdapter.kt

```diff
@@ -53,44 +64,97 @@ class CommunityTournamentPostViewAdapter(
-        if (showTournamentName) {
-            viewHolder.tournamentName.text = tournamentPost.first.name
+        if (showName) {
+            viewHolder.tournamentName.text = post.tournamentName
         } else {
             viewHolder.tournamentName.visibility = View.GONE
         }
 
         viewHolder.imagePath.setImageBitmap(Utils.coordinatesToBitmap(post.run.getPath().getPoints()))
 
-        viewHolder.userName.text = post.userId
+        viewHolder.imagePath.setOnClickListener {
+            showRunDetail(post.run)
+        }
+
+        viewHolder.userName.text = userIdToUsername[post.userId] ?: post.userId
+        if (!userIdToUsername.containsKey(post.userId)) {
+            getUsername(post.userId).whenComplete { it, error ->
+                if (error == null && it != null) {
+                    userIdToUsername[post.userId] = it
+                    viewHolder.userName.text = it
+                }
+            }
+        }
 
         viewHolder.voteCount.text = post.getVotes().toString()
 
+        updateButtonColor(post.getUsersVotes()[userId], viewHolder)
+
         viewHolder.upvote.setOnClickListener {
-            if (post.upvote("user")) {
-                changeButtonColor(viewHolder, it, R.color.red, R.color.grey)
+            if (post.getUsersVotes()[userId] != null && post.getUsersVotes()[userId] == 1) {
+                updateButtonColor(0, viewHolder)
+                viewHolder.voteCount.text = newVoteTotal(post.getVotes(), post.getUsersVotes()[userId], 0).toString()
+                vote(0, post.postId, post.tournamentId)
             } else {
-                changeButtonColor(viewHolder, it, R.color.grey, R.color.grey)
+                updateButtonColor(1, viewHolder)
+                viewHolder.voteCount.text = newVoteTotal(post.getVotes(), post.getUsersVotes()[userId], 1).toString()
+                vote(1, post.postId, post.tournamentId)
             }
-            viewHolder.voteCount.text = post.getVotes().toString()
         }
 
         viewHolder.downvote.setOnClickListener {
-            if (post.downvote("user")) {
-                changeButtonColor(viewHolder, it, R.color.grey, R.color.blue)
+            if (post.getUsersVotes()[userId] != null && post.getUsersVotes()[userId] == -1) {
+                updateButtonColor(0, viewHolder)
+                viewHolder.voteCount.text = newVoteTotal(post.getVotes(), post.getUsersVotes()[userId], 0).toString()
+                vote(0, post.postId, post.tournamentId)
             } else {
-                changeButtonColor(viewHolder, it, R.color.grey, R.color.grey)
+                updateButtonColor(-1, viewHolder)
+                viewHolder.voteCount.text = newVoteTotal(post.getVotes(), post.getUsersVotes()[userId], -1).toString()
+                vote(-1, post.postId, post.tournamentId)
             }
-
-            viewHolder.voteCount.text = post.getVotes().toString()
         }

```

Comment:
Maybe you can add documentation for this part of code.

Reply:
I added a function to modularize the code and some comments